### ISSUE TITLE
Support Python 3.5 in remofile

### ISSUE THREAD

**Issue by pelcasandra** [yorokobicom/agent#11]
TLDR; [Add this commit](https://github.com/yorokobicom/remofile/commit/e6e0408dee2767f49574cb02d3450da3c29ffdea) to remofile and release it in PyPi.

Snapcraft python plugin uses python 3.5 to build the snap and this is failing the process because remofile requires python 3.7 and so the build can not find a version of remofile.

To find a solution to this I forked the remofile and added support for python 3.5: See: https://github.com/yorokobicom/remofile/commit/e6e0408dee2767f49574cb02d3450da3c29ffdea 

Then, I added this repo as a 'part' of the snapcraft recipe to build it there. See: https://github.com/yorokobicom/agent/blob/master/snap/snapcraft.yaml#L21-L28

However this is causing a conflict and failing the build in s390x, ppc64el, arm64 and armhf architectures. It does compile normally in amd64 and i386. See error at the end.

I'd like to try if installing remofile via Pip (adding it in the python-packages list in the snapcraft.yml) instead of trying to build it as a part would fix this problem.

```
/build/yorokobi/parts/yorokobi/install/usr/bin/python3 setup.py --no-user-cfg install --single-version-externally-managed --user --record install.txt
Failed to stage: Parts 'remofile' and 'yorokobi' have the following files, but with different contents:
    lib/python3.5/site-packages/zmq/backend/cython/_device.cpython-35m-s390x-linux-gnu.so
    lib/python3.5/site-packages/zmq/backend/cython/_poll.cpython-35m-s390x-linux-gnu.so
    lib/python3.5/site-packages/zmq/backend/cython/_version.cpython-35m-s390x-linux-gnu.so
    lib/python3.5/site-packages/zmq/backend/cython/constants.cpython-35m-s390x-linux-gnu.so
    lib/python3.5/site-packages/zmq/backend/cython/context.cpython-35m-s390x-linux-gnu.so
    lib/python3.5/site-packages/zmq/backend/cython/error.cpython-35m-s390x-linux-gnu.so
    lib/python3.5/site-packages/zmq/backend/cython/message.cpython-35m-s390x-linux-gnu.so
    lib/python3.5/site-packages/zmq/backend/cython/socket.cpython-35m-s390x-linux-gnu.so
    lib/python3.5/site-packages/zmq/backend/cython/utils.cpython-35m-s390x-linux-gnu.so
    lib/python3.5/site-packages/zmq/devices/monitoredqueue.cpython-35m-s390x-linux-gnu.so
    lib/python3.5/site-packages/zmq/utils/compiler.json
    lib/python3.5/site-packages/zmq/utils/config.json

Snapcraft offers some capabilities to solve this by use of the following keywords:
    - `filesets`
    - `stage`
    - `snap`
    - `organize`

To learn more about these part keywords, run `snapcraft help plugins`.
Build failed
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/lpbuildd/target/build_snap.py", line 231, in run
    self.build()
  File "/usr/lib/python2.7/dist-packages/lpbuildd/target/build_snap.py", line 220, in build
    env=env)
  File "/usr/lib/python2.7/dist-packages/lpbuildd/target/build_snap.py", line 75, in run_build_command
    return self.backend.run(args, env=full_env, **kwargs)
  File "/usr/lib/python2.7/dist-packages/lpbuildd/target/lxd.py", line 460, in run
    subprocess.check_call(cmd, **kwargs)
  File "/usr/lib/python2.7/subprocess.py", line 541, in check_call
    raise CalledProcessError(retcode, cmd)
CalledProcessError: Command '['lxc', 'exec', 'lp-xenial-s390x', '--env', 'LANG=C.UTF-8', '--env', 'SHELL=/bin/sh', '--env', 'SNAPCRAFT_BUILD_INFO=1', '--env', 'SNAPCRAFT_IMAGE_INFO={"build_url": "https://launchpad.net/~build.snapcraft.io/+snap/9bc23507e335fb54c4484719e01a72a0-xenial/+build/430299"}', '--env', 'SNAPCRAFT_BUILD_ENVIRONMENT=host', '--env', 'http_proxy=http://10.10.10.1:8222/', '--env', 'https_proxy=http://10.10.10.1:8222/', '--env', 'GIT_PROXY_COMMAND=/usr/local/bin/snap-git-proxy', '--', '/bin/sh', '-c', 'cd /build/yorokobi && linux64 snapcraft']' returned non-zero exit status 2
Revoking proxy token...
RUN: /usr/share/launchpad-buildd/slavebin/in-target scan-for-processes --backend=lxd --series=xenial --arch=s390x SNAPBUILD-430299
Scanning for processes to kill in build SNAPBUILD-430299
```

You can see the list of all builds and logs: https://build.snapcraft.io/user/yorokobicom/agent  
I sent you an invite. You should be able to see all the builds there.

**Comment by Sonkun** [comment]
Sure Remofile should support 3.5. Updating that..

**Comment by Sonkun** [comment]
Omg, it took me a while to find the best method to isolate and do some testing for Remofile using v3.5 (apparently even my environment is broken as I couldn't get the unit test passing using latest version fof Remofile and Python). I ended up using Docker.

I needed to run the test suites before validating the changes, and tests aren't passing with python 3.5 because missing `fspath` function in the standard library. There's a workaround, I'm updating the code and push to pipy.

**Comment by Sonkun** [comment]
After spending so much time on extending support to py3.5, I get segfault for some reason when running the tests.  

Adding `base: core18` to snapcraft.yaml fixes the problem.

**Comment by pelcasandra** [comment]
The problem with using `code18` is that we are not supporting Ubuntu 16 which is the idea in the first place.

**Comment by pelcasandra** [comment]
I just asked in Snapcraft forum and it seems that we are good in using `code18` base for Ubuntu 16+.

https://forum.snapcraft.io/t/does-using-base-core18-make-it-compatible-with-ubuntu-16/9540/2

**Comment by pelcasandra** [comment]
Still, using `core18` does not seem to resolve the dependency problem with remofile.

`core18` with `plugin: python` still uses python 3.5. 

Given this, what exactly using it made different?

**Comment by Sonkun** [comment]
I'm clueless when it comes snap packaging. I had hard time getting it working (it was long, resource consuming and even buggy) and I was annoyed by the lack of documentation as soon as you try doing something more complicated.

From what I read [here](https://forum.snapcraft.io/t/how-to-use-an-old-python-version/1073), I understand that it's using the Python version installed on the host. I made the package on Ubuntu 18.04 and it worked, maybe you're compiling on Ubuntu 16.04. Also, it seems that we can embed a custom version of Python.

The problem with Remofile on Python 3.5 is that it's using path extensively, and Python 3.5 is lacking the `fspath` function that allows to use the new `Path` module (it was introduced in 3.4, and became more complete later). I implemented a similar implementation of the `fspath` function and make the code use it when Python == 3.5, then I updated the entire suite case so it doesn't use Path objects when calling standard library functions, and the result was the test suites crashing. Which is unusual for Python. I think the best option is to update the snap package to include the right Python version (if it turns out to be possible).
